gubernator/github/classifier.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2016 The Kubernetes Authors All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import json
+
+import models
+
+
+def classify_issue(repo, number):
+    ancestor = models.GithubResource.make_key(repo, number)
+    events = list(models.GithubWebhookRaw.query(ancestor=ancestor))
+    events.sort(key=lambda e: e.timestamp)
+    logging.debug('classifying %s %s (%d events)' % (repo, number, len(events)))
+    return classify(events)
+
+
+def classify(events):
+    '''
+    Given an event-stream for an issue and status-getter, process
+    the events and determine what action should be taken, if any.
+
+    Args:
+        events: a list of GithubWebhookRaw objects
+        status_getter: a function(repo, sha) that returns [GHStatus].
+    Returns:
+        is_pr: bool
+        is_open: bool
+        involved: list of strings representing usernames involved
+        payload: a dictionary of additional information, including:
+            {
+                'author': str author_name,
+                'title': str issue title,
+                'labels': {label_name: label_color},
+                'attn': {user_name: reason},
+                'mergeable': bool,
+                'comments': [{'user': str name, 'comment': comment, 'timestamp': str iso8601}]
+            }
+        last_event_timestamp: the timestamp of the last event that was considered.
+    '''
+    merged = {}
+    labels = []
```

Comment:
It's pretty confusing to switch the type of this from a list to a dict halfway through the function :P